### ISSUE TITLE
List of future improvements

### ISSUE THREAD

**Issue by postlund** [rospogrigio/localtuya#15]
I think it's a good idea to make a list of features and improvements we can make in the future. I'll start with my list:

* Extract code and create base class for entities, e.g. `LocalTuyaEntity` based on a config entry.
* Update entity on-the-fly when options are changed without having to restart Home Assistant (depends first bullet and #14).
* Scanning support in config flow.
* Use scanning to find a device before connecting to it, to handle dynamic IP address changes. I use a similar scheme in the Apple TV component.
* Break out status polling from individual entities and inform entities about updates via the signal interface. This allows us to update devices with multiple entities, e.g. double gang switches, with one request instead of one per entity.
* Make our pytuya implementation actually support the protocol, e.g. by inserting correct sequence number (and other stuff that needs to be handled). The outcome I'm after is to have the network connection up at all time instead of connecting, sending command and immediately disconnect when done. I believe this will be a lot more stable and faster.
* Convert pytuya to asyncio (and the rest of the integration as well).
* Add generic `sensor` and `binary_sensor` platforms that can read any DPS value and present as state.
* Easy set up of helper sensor, e.g. `integration` sensor that calculates `kWh` for used energy.
* General code clean up, removing commented code (it's still available via git history), remove print statements and change some to debug, run `black`, `pylint` and `flake8` on all code.
* Improve documentation, translations in config flows, etc.

Anything I have missed? After some discussion we could make a priority list, create issues for things to do and come up with a roadmap (as supported here on GitHub).

**Comment by andrewmeepos** [comment]
There's been discuss over at https://github.com/rospogrigio/localtuya-homeassistant/pull/10 about adding a "Hacked Position" that lets you set the "position" for devices that don't technically support setting device position. This could be added as an experimental feature or something along those lines. Or is there a better way to approach this? All of my cover devices support set position, so i'm probably not the best person to work on this feature.

**Comment by rospogrigio** [comment]
Somebody on the community thread asked for the climate platform to be introduced.

**Comment by postlund** [comment]
The `climate` platform should be quite easy to implement now if someone knows how what the DPS are used.

**Comment by rospogrigio** [comment]
Let's update the list:
 
DONE: Extract code and create base class for entities, e.g. `LocalTuyaEntity` based on a config entry.
DONE: Update entity on-the-fly when options are changed without having to restart Home Assistant (depends first bullet and #14).
DONE: Scanning support in config flow.
* Use scanning to find a device before connecting to it, to handle dynamic IP address changes. I use a similar scheme in the Apple TV component.
ALMOST DONE: Break out status polling from individual entities and inform entities about updates via the signal interface. This allows us to update devices with multiple entities, e.g. double gang switches, with one request instead of one per entity.
IN PROGRESS: Make our pytuya implementation actually support the protocol, e.g. by inserting correct sequence number (and other stuff that needs to be handled). The outcome I'm after is to have the network connection up at all time instead of connecting, sending command and immediately disconnect when done. I believe this will be a lot more stable and faster.
IN PROGRESS: Convert pytuya to asyncio (and the rest of the integration as well).
DONE: Add generic `sensor` and `binary_sensor` platforms that can read any DPS value and present as state.
* Easy set up of helper sensor, e.g. `integration` sensor that calculates `kWh` for used energy.
DONE: General code clean up, removing commented code (it's still available via git history), remove print statements and change some to debug, run `black`, `pylint` and `flake8` on all code.
* Improve documentation, translations in config flows, etc.
* Positioning handling for covers
* Support to `climate` platform
* Improve support to lightbulbs

**Comment by postlund** [comment]
We should create issues for each remaining thing and perhaps start to use milestones. It makes it easier to plan our work.

**Comment by sajid2310** [comment]
Support for Multi gang switch with FAN.

**Comment by postlund** [comment]
> Support for Multi gang switch with FAN.

Per se already supported, but the fan platform needs some attention.

**Comment by RobyDomo** [comment]
Hi, 
do you think is it possible to have a vacuum platform?
I have a robot that works with smart life app. I already extracted ID and Key, but without vacuum platform i don't think it can work locally. With tuya integration HA has imported the scene, so i can use vaccuum but anyway i use tuya cloud.

**Comment by postlund** [comment]
Probably possible, but they are quite complex so likely requires a bit of work. Maybe you can create a new issue (as a feature request) and post a DPS dump?

**Comment by RobyDomo** [comment]
thank you for you answer @postlund . Sure i can create a new issue as a request, but i don't know how to create a DPS dump.
Thanks for your help. :-)

**Comment by rospogrigio** [comment]
@postlund I can take care of this, looks like MUCH easier than climate platform.
@RobyDomo in order to have a DPS dump you can use the script in tuyadebug.tgz or wait some hours because we are about to have this PR available: https://github.com/rospogrigio/localtuya/pull/79 (or maybe you can just try it and provide feedbacks, actually).

**Comment by postlund** [comment]
@rospogrigio Maybe add some instructions for tuyadebug to readme as I did with debug logs? Or maybe we should move as much as possible over to the wiki instead as that is more flexible?

**Comment by rospogrigio** [comment]
Yes, I agree...

**Comment by rospogrigio** [comment]
@RobyDomo I created the PR #111 for the vacuum implementation. Please post the DPS dump there once you have it, so we don't flood this topic.
Thank you

**Comment by rospogrigio** [comment]
Updated list:
 
* **DONE:** Extract code and create base class for entities, e.g. `LocalTuyaEntity` based on a config entry.
* **DONE:** Update entity on-the-fly when options are changed without having to restart Home Assistant (depends first bullet and #14).
* **DONE:** Scanning support in config flow.
* **DONE:** Use scanning to find a device before connecting to it, to handle dynamic IP address changes. I use a similar scheme in the Apple TV component.
* **DONE:** Break out status polling from individual entities and inform entities about updates via the signal interface. This allows us to update devices with multiple entities, e.g. double gang switches, with one request instead of one per entity.
* **DONE:** Make our pytuya implementation actually support the protocol, e.g. by inserting correct sequence number (and other stuff that needs to be handled). The outcome I'm after is to have the network connection up at all time instead of connecting, sending command and immediately disconnect when done. I believe this will be a lot more stable and faster.
* **DONE:** Convert pytuya to asyncio (and the rest of the integration as well).
* **DONE:** Add generic `sensor` and `binary_sensor` platforms that can read any DPS value and present as state.
* **DONE:** General code clean up, removing commented code (it's still available via git history), remove print statements and change some to debug, run `black`, `pylint` and `flake8` on all code.
* **DONE:** Positioning handling for covers

* IN PROGRESS: Improve documentation, translations in config flows, etc.
* IN PROGRESS: Registration among default HACS integrations
* IN PROGRESS: Improve support to lightbulbs
* IN PROGRESS: Support to `climate` platform
* IN PROGRESS: Support to `vacuum` platform

* Easy set up of helper sensor, e.g. `integration` sensor that calculates `kWh` for used energy.

What an incredible roadmap!

**Comment by postlund** [comment]
An incredible journey indeed!

**Comment by WizBangCrash** [comment]
Great journey indeed and you didn't even include "IN PROGRESS: Support for passive door/window sensors" :-)

**Comment by postlund** [comment]
Ok, that part might not be totally true 😉 But I'm getting closer. Been prioritizing bug reports lately. But will get back on that horse soon.

**Comment by Vismuthxxx** [comment]
Good time of day! how do I add a ZigBee wireless gateway? there is not a single line there is no gateway. Only binary_sensor, cover, fan, light, sensor and switch

**Comment by postlund** [comment]
@Vismuthxxx Not possible since it's not supported yet. There's already a request for it: #69

**Comment by postlund** [comment]
I will go on and close this issue now as I feel that it has served its purpose: give us a start. We have plenty of feature requests already and bugs to deal with. So no need for this.

**Comment by glorifiedg** [comment]
Any plans for IR controllers?

**Comment by postlund** [comment]
@glorifiedg This issue is closed now, but somewhat related to #159?

**Comment by galtamur** [comment]
Hello, I am experiencing this situation: I configured TuyaLocal component with four smartPlugs with current monitoring, and i implemented the template sensors to show the consumption in Lovelace.
It works fine, but from time to time, it looks like the data are not updated (like freezed); when I open on my iphone the SmarLife app and i go into the consumption page, I see number updated, and the are updated also in HA.
Did anyone else experienced the same problem? Any suggestion?
Thank you

**Comment by postlund** [comment]
> Hello, I am experiencing this situation: I configured TuyaLocal component with four smartPlugs with current monitoring, and i implemented the template sensors to show the consumption in Lovelace.
> It works fine, but from time to time, it looks like the data are not updated (like freezed); when I open on my iphone the SmarLife app and i go into the consumption page, I see number updated, and the are updated also in HA.
> Did anyone else experienced the same problem? Any suggestion?
> Thank you

Here: #87

**Comment by hreedijk** [comment]
Hello,

I'm considering to use Local Tuya when the climate part will be released (if you need testers, i'm in !).
Are there plans to implement camera's as well?
I cannot retrieve the feed url from the camera , even the tool ispyconnect cannot find it either, hence the question.

Kind regards,
Harold

**Comment by geckom** [comment]
> The `climate` platform should be quite easy to implement now if someone knows how what the DPS are used.

Hope this might help - tuya climate device.

 *     dps: {
 *         '1': true,  -- power
 *         '2': 23,   -- target temp
 *         '3': 25,   -- current temp
 *         '4': 'DRY',  --- mode  DRY/HEAT/COOLOFF
 *         '5': '1',   - 1 = low, 2 = medium. 3=high (Fan)
 *         '19': 'C',  C or F
 *         '102': false,
 *         '103': 0,
 *         '104': false,  --Swing
 *         '105': 59,
 *         '106': 0,
 *         '110': '0'
 *    }

**Comment by jmargallo** [comment]
Hello
Do you think the Support to Wifi Door Sensor platform will be soon?

I just installed this sensor yours () and even manually setting the deviceid and the local key ... I can't integrate it.
The rest of my devices yours without problem and very grateful

**Comment by realbuz** [comment]
my SP20 (teckin smart wifi-plug) or my outdoor 3 outlet wifi plug (also teckin) have two options per outlet.  One of them is standard on/off as expected, the second one is a countdown timer (`countdown_1 | Integer | {"unit":"s","min":0,"max":86400,"scale":0,"step":1}`

This countdown will toggle (off if on, on if off) after it has rundown to 0.

I dont know how I can currently integrate it in a usable way.  This would give the option to set the timer that does not rely on homeassistant and it's ressources to do it's job).

on my 3 outlet switch I have countdown_1, _2, _3

thanks

**Comment by txespirito** [comment]
I bought a switch with a lud led but I can't set the light to 100%. The dps indicated are the following that do not correspond to the lights indicated in the documentation.

    {
      "activeTime": 1614962379,
      "devId": "xxxxxxxxxxxxxxxxxxxxxxx",
      "displayOrder": 0,
      "dpMaxTime": 1614963543272,
      "dpName": {
        "1": "Interruptor 1",
        "27": "Led Switch"
      },
      "dps": {
        "1": false,
        "9": 0,
        "27": false,
        "28": "colour",
        "31": "ff00100164ffff",
        "32": "3855b40168ffff",
        "33": "ffff320100ff00",
        "34": "ffff5003ff000000ff000000ff000000000000000000",
        "35": "ffff3201ff0000",
        "36": "ffff0505ff000000ff00ffff00ff00ff0000ff000000",
        "37": 0
      },

Has anyone managed to configure it correctly and could update the documentation?

Thanks

This is the link of the switch/ligth model:
https://es.aliexpress.com/item/1005001681604103.html?spm=a2g0s.9042311.0.0.365763c0lUQCEj

**Comment by fabiohblage** [comment]
Hello how to add a four button switch.

**Comment by bfmorgan** [comment]
I have a Feit dimmer switch. It would be nice if dimmers were added to the list of devices supported. If there is a yaml way of integrating them now, can someone lead me in that direction? thanks.

**Comment by sercri** [comment]
The scan function for those ip that dinamycally changed it is not working any more since core-2021.4.6 update on an hassio setup rb3+.
I cannot set a static ip because the 4g modem does not allow it; it used to work good before the last hassio update - so that now I must  change the ip on the config file for this device, every time it goes offline.

Thankyou for yur effort

**Comment by javamahn** [comment]
Any possibly of adding additional network options for autodiscovery? I have my HomeAssistant on my main network but my IoT are on a seperate Guest network

**Comment by jonniepanda** [comment]
Hi Guys - awesome work thanks for sharing this. Any news on the climate platform?

**Comment by malpeli** [comment]
How about an option to use HA's media player as a defacto microphone for music mode?

**Comment by ChrisRonaasen** [comment]
**This issue is closed. Open a new issue or make a feature request (after checking that it hasn't already been mentioned) if you want to contribute or ask questions.**



_soooohowaboutthosedoorsensorsandcameras?_

**Comment by cloudbr34k84** [comment]
hi there is there any chance of getting Dc inverter Swimming Pool Heat Pump. I can get all the information from Tuya Platform, and i can connect but i dont think this device is supported? Any help

**Comment by JohanTess** [comment]
Tuya Cat Litter Box Wifi (round model Wifi)
Found following values
101 = on/off (device on/off) -  switch
102 = Trigger manual cleaning - switch (how do I select a toggle switch?)
103 = Delay before start cleaning - value between 60 and 1800 - number (time in seconds)
104 = ?
105 = UV disinfection time - value between 60 and 1800 - number (time in seconds)
106 = ? value true/false but didn't find what trigger this state
107 = ? 1200 didn't find anything that changes this value
108 = LED ring on/off - switch
109 = Sleep mode on/off - switch
110 = Start time of sleep mode - 24h time format - 0 = 00;00, 60 = 01:00,...., 1230 = 20:30
111 = End time of sleep mode (same as 110 as format)
112 = ? value is showing 1
113 = ? value is showing 25
114 = Status of the device following values discovered Standby, Cleaning, Caking, Power Off (how to get the values is lovelace?)
115 = UV disinfection on/off - switch
116 = ? set to false
117 = Cat sand level - Values found Empty, Half, Full - (how to get the values is lovelace?) After filling up you need to push cat cand reset on the app...probably one of the ? is the trigger
118 = Error - 0 = no error, 2 = tray removed... unable to find other error codes
119 = N/A
120 = ? set to False

If anyone can give me some clues on what platform to use for the above and how to configure.... THX in advance.

**Comment by boheme61** [comment]
Nedis WIFIPO120FWT plug with power-consupmtion, have 14 different DP''s ... how do i know which is what, i did get a "spec" out of iot.tuya.com, but doesnt tell everything, it only specifies 8 out of 14 ... but my problem is more what kind of "Platform" i should choose for the different ID's, i know initially "Switch" next ID is countdown_1 ( i asume switch ?) , next 4 ID's are most likely sensors (power, voltage, cur_current, add_ele) then i have ( cycle_time and random_time ), what "platform" to choose here ? , number, switch, select ? ... and then i have the rest, howto know what they are good for ? 
.............................................................................
"product_name": "Outdoor Smart Plug",
    "status": [
"status": [
        "code": "switch_1",
        "dp_id": 1,
        "type": "Boolean",
        "values": "{}"

        "code": "countdown_1",
        "dp_id": 9,
        "type": "Integer",
        "values": "{\"unit\":\"s\",\"min\":0,\"max\":86400,\"scale\":0,\"step\":1}"

        "code": "add_ele",
        "dp_id": 17,
        "type": "Integer",
        "values": "{\"unit\":\"kwh\",\"min\":0,\"max\":50000,\"scale\":3,\"step\":100}"

        "code": "cur_current",
        "dp_id": 18,
        "type": "Integer",
        "values": "{\"unit\":\"mA\",\"min\":0,\"max\":30000,\"scale\":0,\"step\":1}"

        "code": "cur_power",
        "dp_id": 19,
        "type": "Integer",
        "values": "{\"unit\":\"W\",\"min\":0,\"max\":50000,\"scale\":1,\"step\":1}"

        "code": "cur_voltage",
        "dp_id": 20,
        "type": "Integer",
        "values": "{\"unit\":\"V\",\"min\":0,\"max\":5000,\"scale\":1,\"step\":1}"

        "code": "cycle_time",
        "dp_id": 41,
        "type": "String",
        "values": "{}"

        "code": "random_time",
        "dp_id": 42,
        "type": "String",
        "values": "{}"
.............................................................................
Beside above i have
 "dp_id": 21 , 22 , 23 , 24 , 25 , 26

would appreciate help understanding which "Platform " to choose above

Future Suggestion / Improvement
Would be nice to be able to "regret" (going back) in the gui, so old men who cant keep count, don't click add another entity, just to find out that i already configured all 14, and end up with 1 choice , the infamous X in the top right corner :(

**Comment by lbijzeit** [comment]
After days of trying to get things in home assitant I now notice that there isn't support yet for led controllers (stand alone) and for the MiBoxer WL-Box1  (2.4 Ghz Gateway). It would be fantastic if this would be supported by local tuya.
Could it be put on the list of improvements?

**Comment by marcokreeft87** [comment]
First of all, I love the integration! But the climate ( in this case an portable airco ) isnt working like the tuya integration.

![image](https://user-images.githubusercontent.com/10223677/180382184-f579fae3-bea0-425c-a225-91f8c1d6f554.png)

**Comment by cloudbr34k84** [comment]
Is possible to add a new sensor to an already added device or do you have delete and start again?

**Comment by rospogrigio** [comment]
You need to delete the device and re-add it from scratch, sorry.

**Comment by cloudbr34k84** [comment]
> You need to delete the device and re-add it from scratch, sorry.

Feature request. Lol

**Comment by MajaGithuab** [comment]
after everything was working super fine I started having devices unavailable, IPs are fixed for all Tuya devices, it is showing online on the Tuya developer web page, I can control it through Tuya mobile App.
I removed devices and add it again, removed the integration totally and added it again, it works for a while then it becomes unavailable.

Any advice ? I dont want to return to the Tuya cloud integration.

**Comment by guiseco** [comment]
> You need to delete the device and re-add it from scratch, sorry.

Sorry, but wouldn't it be easier to be able to directly tweak the configuration file for a more precise adjustment?

**Comment by FS1961** [comment]
Can you please add a way to save the configuration of the Tuya devices, like an export list function or something. I've had to rebuild this a few times from scratch so it would be nice to save the current config to a file with all the info. Thanks!

**Comment by cloudbr34k84** [comment]
Hey I tried to start a LocalTuya DPS thread in HA community. Feel free to add yours there for now so everyone can benefit.
https://community.home-assistant.io/t/localtuya-device-dp-config-settings-thread/458054/2?u=cloudbr34k

**Comment by theundefined** [comment]
I think it would be great to combine many different things in one place like documentation ;)
I've prepared list for one my device (LSC  Smart LED Flood light with sensor) on https://docs.google.com/document/d/1nv6GIfD81kXIyWxQ4I_gRBDBjMrCBfT3V7nIK2CuBqo/edit?usp=sharing

**Comment by jmcollin78** [comment]
> 

How do you delete a device with the LocalTuya integration (there is no delete option) ? Removing all DP is bad because you cannot get it back.

**Comment by rospogrigio** [comment]
> How do you delete a device with the LocalTuya integration (there is no delete option) ? Removing all DP is bad because you cannot get it back.

They have hidden the button but it's still there. You enter Integrations -> select the desired device -> 3 dots beside the Download Diagnostics button -> Delete.

**Comment by FS1961** [comment]
Just go to "Devices and settings", devices, type the device, hit the three dots next to "download diagnostics" and a drop down will show up to delete the device.

**Comment by georgebancila** [comment]
Hello, is there a way to add a new entity to an existing device? Thanks

**Comment by FS1961** [comment]
Edit the device is an option.

**Comment by georgebancila** [comment]
I was asking about adding a new entity in that edit device screen. But I just saw in other comments that it is not possible

**Comment by Din-BH** [comment]
One of my devices that has been working for quite some time just stopped working and entities from it lists as unavailable. I used to have it installed via a Node-red component as well, and that has stopped working too - the integration there still connects to teh device, but data requested from it comes up garbled like the data format would have been changed. Could the protocol have been upgraded OTA? Anyone encountered something similar?

**Comment by 1st-BrainStormer** [comment]
It would be nice to be able to transfer devices from the Tuya Cloud Integration so we don't have to manually put in all the data points on each device when switching over.

**Comment by increddibelly** [comment]
I totally agree with Brainstormer here. It's the only thing that's holding me back; the hassle of doing the requests that could be automated.
using the api call at https://eu.iot.tuya.com/micro-app/cloud/api/v5_1/project/device/list will give all device Id's
enumerating those through the device details api https://eu.iot.tuya.com/cloud/explorer?id=<<account id>>&groupId=group-controll&interfaceId=<<device id>> should be doable, right?
and while I was digging for these, I found https://openapi.tuyaeu.com/v1.0/token?grant_type=1&terminal_id=11 gets a security token.

**Comment by MrHaroldA** [comment]
> Sorry, but wouldn't it be easier to be able to directly tweak the configuration file for a more precise adjustment?

Is that still possible? I'd like to set the min/max range of a climate entity created by Local Tuya, but I can't find any yaml with the entity settings ...

**Comment by Bohr80** [comment]
I agree with BrainStormer, it would be good be get all devices from Tuya IoT Platform. 
At the moment setting up Tuya IoT Platform doesn't work - gives error 1106: permission deny. 
At the same moment Tuya integration using the same IDs and passwords works perfectly fine.

**Comment by cloudbr34k84** [comment]
> I agree with BrainStormer, it would be good be get all devices from Tuya IoT Platform. 
> At the moment setting up Tuya IoT Platform doesn't work - gives error 1106: permission deny. 
> At the same moment Tuya integration using the same IDs and passwords works perfectly fine. 

That's a different issue. Go to HA community and search for that error code. There are lots of discussions on how to fix that issue. 👍

**Comment by 1st-BrainStormer** [comment]
> I agree with BrainStormer, it would be good be get all devices from Tuya IoT Platform. At the moment setting up Tuya IoT Platform doesn't work - gives error 1106: permission deny. At the same moment Tuya integration using the same IDs and passwords works perfectly fine.

That error is that the free month access is up and you need to extend the test period.

**Comment by Bohr80** [comment]
> > I agree with BrainStormer, it would be good be get all devices from Tuya IoT Platform. At the moment setting up Tuya IoT Platform doesn't work - gives error 1106: permission deny. At the same moment Tuya integration using the same IDs and passwords works perfectly fine.
> 
> That error is that the free month access is up and you need to extend the test period.

I don't think so, because it hasn't been a month yet since I've set my account on Tuya IoT Platform. 
But thanks for that suggestion, I would probably wonder what is going on after a month. ;)

**Comment by avenger11** [comment]
Please add the support for Humidifier on the TODO list :)

**Comment by rtech73** [comment]
Is there any plan to add support for their LED light strips?    I just got one with TV sync and while the device is found and integrated, it indicates “not supported”.

**Comment by jonwestuk** [comment]
> Please add the support for Humidifier on the TODO list :)

and if possible dehumidifiers - with the ability to set humidity level on the device, activate childlock, fan speed etc.

**Comment by abcben78** [comment]
Just install the latest update, try to install a sensor and it hang the whole Local Tuya Integration. Nothing works anymore, it was already strange that it not discovering anymore.

**Comment by rospogrigio** [comment]
Well that's really strange, I have received many positive feedbacks and I am using the new version myself with no issues at all.
Did you try restarting HA and/or reverting to the previous version of localtuya? Have you also tried to enable debugging (checking the new debug button) when adding the new sensor? Can you share the log messages?
Thank you

**Comment by abcben78** [comment]
Hello, hereby the logs as requested, hope that you do something with it.

The first thing what I saw was, that the discovery wasn't showing anything
anymore, so I'm using the ... for adding a motion sensor and after that I
have got the error. I find out that in .storage directory 3 core files are
changed so I restore them from my backup after this I hat all after 3.5
hours online again... 😅

Since I use LocalTuya, I have no illumination sensors,  motion sensors, and
door open sensors anymore.

Motion sensors and contact sensors can only be installed as sensor and not
as binary sensor, the result is a very slow sensor that can't be used for
the lights anymore, only for a late triggering the alarm. In Tuya OK

Illumination sensors can't be installed at all. In Tuya OK

The 4" Terminal can't be installed at all. In Tuya OK

The coffee machine can't be installed. In Tuya OK

The buttons (Zibee) can't be installed. In Tuya OK

The sockets (Wifi and Zigbee) are working fine, but the energy, power, etc.
can I only been seeing as a attribute. In Tuya OK

The switches  are  working fine, but timers, back-light, etc. are not there
anymore. In Tuya OK

The bulbs work fine, there is nothing what I can find that doesn't work. In
Tuya OK

The shutters are working fine, there is nothing what I can find that
doesn't work. In Tuya OK

When you need more information I'm glad to give it.

Yes, I can have some help here 😁😁😁

You need a beta tester?

Greetings,

Ben

**Comment by rospogrigio** [comment]
@abcben78 well, I had a lot of users who tested and they all gave positive feedbacks, so I guess there's something strange with some of your devices. The changes introduced with 5.0.0 are quite extensive but they don't justify all these failures.
You said you posted the logs but you actually didn't so please retry.
Then here is how I'd proceed:
- I'd revert to localtuya 4.x.x and check that everything works as before
- then I'd remove the integration entirely, upgrade to 5.0.0 and start re-adding the devices one by one, starting with the most problematic ones, enabling the debug with the new button I added. Once we are able to add the device successfully the debug can be disabled for that device
Send your logs in the meantime so I can check if there is some clear error.
Thank you

**Comment by abcben78** [comment]
I did the same thing as you did and the error didn't occurs anymore. Is it
so that everything what work in Tuya also can work in LocalTuya? Thank you.

**Comment by rospogrigio** [comment]
Well... no. Tuya works via cloud APIs, localtuya uses a local socket to send commands directly. Also, there is a new protocol that Tuya is using that seems no-one was able to decrypt yet (maybe protocol 3.5? who knows). Of course everything that works in Tuya COULD work in LocalTuya but it has to be coded differently.

**Comment by rumlak** [comment]
Support for the wifi pet feeders would be a great addition

**Comment by TolianIPB** [comment]
Hi guys! I really want to see support for Zigbee network Getwways by IP in Local Tuya. It would be an awesome solution.
Something like this:
https://github.com/leeyuentuen/localtuya/

Your branch is very positive branch and you fix problems quickly.
Please tell me, is there any hope that Zigbee will appear in your branch?

**Comment by cloudbr34k84** [comment]
the ability to add entities to an already created device. You currently have to start all over again

**Comment by patrik-malina** [comment]
> the ability to add entities to an already created device. You currently have to start all over again

+1

**Comment by cloudbr34k84** [comment]
Number entities are supported but this is a fan. I think there is fan domain you can choose??

**Comment by cloudbr34k84** [comment]
Number entities are supported but this is a fan. I think there is fan domain you can choose?? Try playing with that 

![Screenshot_20230307-230744.png](https://user-images.githubusercontent.com/58960644/223431027-93051506-e0f9-46d4-b6b7-e24029a61327.png)

**Comment by wenstarner** [comment]
Thank you—I tried that, but it doesn’t work!

Even more interesting is what I found after some more testing:

1. The number entity will READ the current fan speed (a number from 1 to 8, inclusive). So when I change the speed on the fan itself, the number entity updates in the logbook.
2. I cannot WRITE a new number to the fan. The slider stays on the number I select, but the logbook shows no actual update. Also, I can see that the fan speed has not changed.
3. An exception to Point 2: When I set the fan speed on the fan itself, adjusting the number entity slide up or down will put the fan into some kind of ‘0’ power state (seems to operate at the same power as Level 1). Strangely, when I go to add a new device while on this mode, Local Tuya reports for this DP ID the value that I manually set on the unit—so the slide changed the fan speed to low, but the unit doesn’t seem to register that it’s changed to the strange ‘0’ lower speed.

Anyway, I suppose long story short is: Does anyone know how I can get the number entity slider to successfully write to the unit? As it currently can read fine.

I’m not to HA so am happy to provide logs, but thought I’d ask this now as switches, etc. work perfectly fine. Thanks!

**Comment by lukiluki123** [comment]
> Hi guys! I really want to see support for Zigbee network Getwways by IP in Local Tuya. It would be an awesome solution. Something like this: https://github.com/leeyuentuen/localtuya/
> 
+1

**Comment by GarfieldMZ** [comment]
Hey!
Is there any chance to get a "Pet Feeder" integrated (like [this](https://expo.tuya.com/product/324064))

Thanks!!

**Comment by MikeLehnitz** [comment]
Hello, I would like temperature/humidity sensors (WIFI) to be integrated.

**Comment by ralvarez83** [comment]
Hi, I would like Smart Bluetooth Soil Tester Temperature and Humidity Meter to be integrated. A product example is [this](https://www.aliexpress.com/item/1005005816297100.html).

Thanks so much

**Comment by sommelierak111** [comment]
Hi. Tuya makes Garage Door Controller CKM01, logicaly its a short time connection relay with binary sensor of opened/closed status. And this device may be used also as controller for electric mechanical locks on the doors. 
But Tuya addon integrates it to HA as "cover".
It is strange to see it in "cover" class, because  it is more "lock" than cover.
It will be great to integrate this device via localtuya as a lock.
Now I only can integrate it as two different devices: 1) button (switch) to control door motor and 2) as binary sensor for status (opened/closed).

**Comment by phelipebf** [comment]
Maybe we can start mapping DPs for some devices. I did this for Inkbird IHC-200 WiFi:

```
Sensors:
104: PV (current humidity)
106: SV (target humidity)

Mode:
115: 
- 1 = Work2 ON (Dehumidification)
- 2 = OFF
- 3 = Work1 ON (Humidification)

Settings:
102: Humidity Calibration
108: Dehumidification Delay Time
109: Alarm High Humidity Limit
110: Alarm Low Humidity Limit
117: Humidification Diferential Value
118: Dehumidification Diferential Value

UNKNOWN:
12: (number)
111: (boolean)
112: (boolean)
113: (boolean)
119: (boolean)
```

**Comment by cloudbr34k84** [comment]
> > Maybe we can start mapping DPs for some devices. I did this for Inkbird IHC-200 WiFi:
> > ```
> > Sensors:
> > 104: PV (current humidity)
> > 106: SV (target humidity)
> > 
> > Mode:
> > 115: 
> > - 1 = Work2 ON (Dehumidification)
> > - 2 = OFF
> > - 3 = Work1 ON (Humidification)
> > 
> > Settings:
> > 102: Humidity Calibration
> > 108: Dehumidification Delay Time
> > 109: Alarm High Humidity Limit
> > 110: Alarm Low Humidity Limit
> > 117: Humidification Diferential Value
> > 118: Dehumidification Diferential Value
> > 
> > UNKNOWN:
> > 12: (number)
> > 111: (boolean)
> > 112: (boolean)
> > 113: (boolean)
> > 119: (boolean)
> > ```
> 
> we need this badly!

How did you get inkbird connected to tuya?? I have the wifi pool thermometer but could get it to connect

**Comment by phelipebf** [comment]
> 
> How did you get inkbird connected to tuya?? I have the wifi pool thermometer but could get it to connect

Pair with Tuya Smart app and it will be available in the cloud through this integration.

**Comment by cloudbr34k84** [comment]
> > How did you get inkbird connected to tuya?? I have the wifi pool thermometer but could get it to connect
> 
> Pair with Tuya Smart app and it will be available in the cloud through this integration.

Mmm tried that.. wouldn't pair

**Comment by MarekOleksik** [comment]
Hi, Is there any chance to support 3.5 Tuya protocol?

**Comment by C74K** [comment]
any chance to support 3.5 Tuya protocol ?
i have led with 3.5 ;/

**Comment by jjgerek** [comment]
Any idea when the 3.5 protocol will be supported? All new Tuya devices now running on 3.5 and they are not supported in Local Tuya

**Comment by phelipebf** [comment]
> Any idea when the 3.5 protocol will be supported? All new Tuya devices now running on 3.5 and they are not supported in Local Tuya

Take a look at that fork: https://github.com/xZetsubou/hass-localtuya/

**Comment by rossens** [comment]
> > Any idea when the 3.5 protocol will be supported? All new Tuya devices now running on 3.5 and they are not supported in Local Tuya
> 
> Take a look at that fork: https://github.com/xZetsubou/hass-localtuya/

I took a stab at it: https://github.com/rospogrigio/localtuya/pull/2178

(Wish I'd seen that fork before hand - might have helped lol)

Update: I just tried the [xZetsubou/hass-localtuya/](https://github.com/xZetsubou/hass-localtuya/) integration and I highly recommend it. These are just my first impressions after a quick play:
- Many problems fixed for my 3.5 LED light and my other 3.3 devices mostly smart plugs (switches).
- Light entity seems much improved (eg scenes fixed)
- Switch entity much improved
- General workflow much improved, adding devices etc

This repo seems not effectively maintained anymore - with 135 open PRs. Most with no reviews, no comments.
